src/plugins/mavlink_ftp/include/plugins/mavlink_ftp/mavlink_ftp.h

```diff
@@ -0,0 +1,189 @@
+#pragma once
+
+#include <vector>
+#include <memory>
+#include <string>
+#include <functional>
+
+// This plugin provides/includes the mavlink 2.0 header files.
+#include "mavlink/v2.0/common/mavlink.h"
+#include "plugin_base.h"
+
+namespace mavsdk {
+
+class System;
+class MavlinkFTPImpl;
+
+/**
+ * @brief The MavlinkFTP class provides MAVLink FTP functionality
+ *
+  */
+class MavlinkFTP : public PluginBase {
+public:
+    /**
+     * @brief Constructor. Creates the plugin for a specific System.
+     *
+     * The plugin is typically created as shown below:
+     *
+     *     ```cpp
+     *     auto mavlink_ftp = std::make_shared<MavlinkFTP>(system);
+     *     ```
+     *
+     * @param system The specific system associated with this plugin.
+     */
+    explicit MavlinkFTP(System &system);
+
+    /**
+     * @brief Destructor (internal use only).
+     */
+    ~MavlinkFTP();
+
+    /**
+     * @brief Possible results returned for requests.
+     */
+    enum Result : uint8_t {
+        SUCCESS,
+        ERR_FAIL,                  ///< Unknown failure
+        ERR_INVALID_DATA_SIZE,     ///< PayloadHeader.size is invalid
+        ERR_INVALID_SESSION,       ///< Session is not currently open
+        ERR_NO_SESSIONS_AVAILABLE, ///< All available Sessions in use
+        ERR_EOF,                   ///< Offset past end of file for List and Read commands
+        ERR_UNKOWN_COMMAND,        ///< Unknown command opcode
+        ERR_FAIL_FILE_EXISTS,      ///< File exists already
+        ERR_FAIL_FILE_PROTECTED,   ///< File is write protected
+
+        // These error codes are returned to client without contacting the server
+        ERR_OP_IN_PROGRESS = 200,  ///< Operation is already in progress
+        ERR_INVALID_PARAMETER,     ///< Invalid parameter
+        ERR_FAIL_LOCAL_OPEN,       ///< Failed to open local file
+        ERR_FAIL_LOCAL_READ,       ///< Failed to read local file
+        ERR_FAIL_LOCAL_WRITE,      ///< Failed to write local file
```

Comment:
I think we should not expose the underlying MAVLink FTP errors but we should wrap the errors that we are actually expecting to something simple that a user implementing against the library can understand.

E.g.
```
CONNECTION_ERROR,
PROTOCOL_ERROR,
UNSUPPORTED,
FILE_EXISTS,
FILE_PROTECTED,
IN_PROGRESS
```
```

Reply:
Implemented translation between ftp server return codes and SDK user return codes